### PR TITLE
Add complex group algorithms

### PR DESCRIPTION
**Pull request by ldrumm**
Add complex group algorithms

Extends the implementation of group algorithms by adding support for
complex types. Operations that can be done element-wise are implemented
in headers and multiplication is implemented in CUDA PI. This involved
changing the macros in CUDA PI to accommodate complex and defining
complex types to use in PI interface. The scratch space for moving data
between workgroups in the PI also had to be doubled.

This patch depends on two fixes to the NVPTX backend which are exposed by the new code generation.

Tests are in [intel/llvm-test-suitegit](https://github.com/ldrumm/llvm-test-suite/tree/complex_group_algo)

### CODE REVIEW

File: sycl/include/CL/__spirv/spirv_ops.hpp [hunk 1]
```diff
@@ -928,6 +929,14 @@ __CLC_BF16_SCAL_VEC(uint32_t)
 #undef __CLC_BF16_SCAL_VEC
 #undef __CLC_BF16
 
+__SYCL_CONVERGENT__ extern SYCL_EXTERNAL __SYCL_EXPORT __spv::complex_half
+__spirv_GroupCMulKHR(unsigned int, unsigned int, __spv::complex_half) noexcept;
+__SYCL_CONVERGENT__ extern SYCL_EXTERNAL __SYCL_EXPORT __spv::complex_float
+__spirv_GroupCMulKHR(unsigned int, unsigned int, __spv::complex_float) noexcept;
+__SYCL_CONVERGENT__ extern SYCL_EXTERNAL __SYCL_EXPORT __spv::complex_double
+__spirv_GroupCMulKHR(unsigned int, unsigned int,
```
Comment (Naghasan) [inline sycl/include/CL/__spirv/spirv_ops.hpp:937]: AFAIK there no KHR extension for these, you should change for `GroupCMulINTEL`

File: libclc/generic/include/spirv/spirv_types.h [hunk 2]
```diff
@@ -46,4 +46,16 @@ enum GroupOperation {
   ExclusiveScan = 2,
 };
 
+typedef struct {
```
Comment (asudarsa) [inline libclc/generic/include/spirv/spirv_types.h:49]: Would it make sense to add bfloat16 here? It is still 'experimental', but will be moved out soon. We can wait for that to happen. Thanks

**Review by romanovvlad — COMMENTED** [review]
LGTM overall, but would like other reviewers to approve

**Comment by ldrumm** [comment]
Thanks for the reviews and apologies for the silence. I left this while dependent patches were merged upstream.

File: libclc/generic/include/spirv/spirv_types.h (hunk 2 above)
Comment (ldrumm) [reply libclc/generic/include/spirv/spirv_types.h:49]: This has already got rather long in the tooth, so I'd like to defer that change for later

File: sycl/include/CL/__spirv/spirv_ops.hpp (hunk 1 above)
Comment (ldrumm) [reply sycl/include/CL/__spirv/spirv_ops.hpp]: Good catch. Fixed

**Review by ldrumm — COMMENTED** [review]
Sorry for the long delay folks. I had to get some other thorny changes merged before this actually works.

I think I've addressed everything so far, so if you have the bandwidth I think it's ready for review

Many thanks!

**Review by Naghasan — APPROVED** [review]
LGTM, I'm not entriely sure `ptx-nvidia` is the right folder for the collectives.cl additions as I don't see anything nvptx specific unless `__clc__get_group_scratch_complex_*` is exclusively meant to be implemented for nvptx. (has been a while for me...)

**Comment by npmiller** [comment]
friendly ping @sergey-semenov @asudarsa @intel/llvm-gatekeepers 

Could we get some help/reviews on this PR, it's been open for a long time.

**Review by sergey-semenov — APPROVED** [review]
Sorry about the late review, LGTM

**Pull request merged**